### DISCUSSION TITLE
Can I use the same texture for render target and shader resource at the same time?

### DISCUSSION THREAD

**Discussion by FlightBlaze** [DiligentGraphics/DiligentEngine#174] [Q&A] [answered]
Can I use the same texture for render target and shader resource at the same time?
I'm using SRV for accessing texture texels in the shader while rendering into the same texture with RTV and I get a lot of errors in my console:

```
Diligent Engine: ERROR: Vulkan debug message (validation): VUID-VkDescriptorImageInfo-imageLayout-00344
                 Validation Error: [ VUID-VkDescriptorImageInfo-imageLayout-00344 ] Object 0: handle = 0x10126f668, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xde55a405 | vkCmdDrawIndexed():

Cannot use VkImage 0x167d5d0000000cb4[Render Target Color] (layer=0 mip=0) with specific layout 
VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL that doesn't match the previous known layout 
VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.

The Vulkan spec states: imageLayout must match the actual VkImageLayout of each subresource accessible from imageView at the time this descriptor is accessed as defined by the image layout matching rules (https://vulkan.lunarg.com/doc/view/1.2.189.0/mac/1.2-extensions/vkspec.html#VUID-VkDescriptorImageInfo-imageLayout-00344)
                 Object[0] (cmd buffer): Handle 0x10126f668
```

```
Diligent Engine: ERROR: Vulkan debug message (validation): VUID-vkCmdDrawIndexed-None-02699
                 Validation Error: [ VUID-vkCmdDrawIndexed-None-02699 ] Object 0: handle = 0x7cff3a0000000cc2, name = Implicit signature of PSO 'blazevg solid color PSO' - static/mutable set, type = VK_OBJECT_TYPE_DESCRIPTOR_SET; | MessageID = 0xa44449d4 | Descriptor set VkDescriptorSet 0x7cff3a0000000cc2[Implicit signature of PSO 'blazevg solid color PSO' - static/mutable set] encountered the following validation error at vkCmdDrawIndexed() time:

Image layout specified at vkUpdateDescriptorSet* or vkCmdPushDescriptorSet* time doesn't match actual image layout
at time descriptor is used. See previous error callback for specific details.
The Vulkan spec states: Descriptors in each bound descriptor set, specified via vkCmdBindDescriptorSets, must be valid if they are statically used by the VkPipeline bound to the pipeline bind point used by this command 

(https://vulkan.lunarg.com/doc/view/1.2.189.0/mac/1.2-extensions/vkspec.html#VUID-vkCmdDrawIndexed-None-02699)
                 Object[0] (descriptor set): Handle 0x7cff3a0000000cc2, Name: 'Implicit signature of PSO 'blazevg solid color PSO' - static/mutable set'
```

```
Diligent Engine: ERROR: Binding render targets (DeviceContextVkImpl::TransitionRenderTargets)
requires texture 'Render Target Color' to be transitioned to RENDER_TARGET state.
Actual texture state: SHADER_RESOURCE. Use appropriate state transition flags or explicitly
transition the texture using IDeviceContext::TransitionResourceStates() method.
```

**Comment by TheMostDiligent** [comment]
This is prohibited in general case, but may be possible in certain specific scenarios.
What are you trying to do?

**Reply by FlightBlaze** [reply]
I want to add blending modes to my 2d graphics library. Therefore, I use the SRV of background to blend the resulting color with the background in the pixel shader using formulas specific to particular blending mode

**Reply by FlightBlaze** [reply]
Now I want to create a Gaussian blur effect behind a translucent object in one draw call. Is it even possible to render a batch of rectangles, that share a single texture and using the backdrop blur effect I described above? I could render a rectangle fully blurred, but I want to blur only a specific part of it, that is being overlapped by a translucent object (or many of them)

You have said that using the same texture for render target and shader resource at the same time may be possible in certain specific scenarios. In which scenarios that's possible?

By the way, in OpenGL that's actually possible and I have even done this blur effect without using another render target

**Reply by TheMostDiligent** [reply]
I am not sure what you are trying to achieve. If you want to render rectangles to a render target, which you also want to blur at the same time, it is not possible. To compute blur, you need to access neighboring pixels that may be overwritten by neighboring pixel shaders, so there will be race condition.

**Comment by TheMostDiligent** [comment, marked as answer]
This usage is invalid. You should define [blend state](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngine/interface/BlendState.h#L376) in the [pipeline state description](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngine/interface/PipelineState.h#L252).

**Reply by FlightBlaze** [reply]
I don't know how to craft the multiplicative blending from SrcBlend, DstBlend and BlendOp.

Formula: Source color * Destination color

Is there any kind of _BLEND_OPERATION_MULTIPLY_?

**Reply by TheMostDiligent** [reply]
[This post](https://learnopengl.com/Advanced-OpenGL/Blending) describes alpha-blending in OpenGL. It is the same in Diligent
